// @paper-design/shaders-react/dist/index.js
/* * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                    Paper Shaders                    *
 *       https://github.com/paper-design/shaders       *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * */

"use client";
import { ShaderMount } from "./shader-mount.js";
import { MeshGradient, meshGradientPresets } from "./shaders/mesh-gradient.js";
import { SmokeRing, smokeRingPresets } from "./shaders/smoke-ring.js";
import { NeuroNoise, neuroNoisePresets } from "./shaders/neuro-noise.js";
import { DotOrbit, dotOrbitPresets } from "./shaders/dot-orbit.js";
import { DotGrid, dotGridPresets } from "./shaders/dot-grid.js";
import { SimplexNoise, simplexNoisePresets } from "./shaders/simplex-noise.js";
import { Metaballs, metaballsPresets } from "./shaders/metaballs.js";
import { Waves, wavesPresets } from "./shaders/waves.js";
import { PerlinNoise, perlinNoisePresets } from "./shaders/perlin-noise.js";
import { Voronoi, voronoiPresets } from "./shaders/voronoi.js";
import { Warp, warpPresets } from "./shaders/warp.js";
import { GodRays, godRaysPresets } from "./shaders/god-rays.js";
import { Spiral, spiralPresets } from "./shaders/spiral.js";
import { Swirl, swirlPresets } from "./shaders/swirl.js";
import { Dithering, ditheringPresets } from "./shaders/dithering.js";
import { GrainGradient, grainGradientPresets } from "./shaders/grain-gradient.js";
import { PulsingBorder, pulsingBorderPresets } from "./shaders/pulsing-border.js";
import { ColorPanels, colorPanelsPresets } from "./shaders/color-panels.js";
import { StaticMeshGradient, staticMeshGradientPresets } from "./shaders/static-mesh-gradient.js";
import { StaticRadialGradient, staticRadialGradientPresets } from "./shaders/static-radial-gradient.js";
import { PaperTexture, paperTexturePresets } from "./shaders/paper-texture.js";
import { FlutedGlass, flutedGlassPresets } from "./shaders/fluted-glass.js";
import {} from "./shaders/fluted-glass.js";
import {} from "@paper-design/shaders";
import { Water, waterPresets } from "./shaders/water.js";
import { ImageDithering, imageDitheringPresets } from "./shaders/image-dithering.js";
import { Heatmap, heatmapPresets } from "./shaders/heatmap.js";
import { LiquidMetal, liquidMetalPresets } from "./shaders/liquid-metal.js";
import { HalftoneDots, halftoneDotsPresets } from "./shaders/halftone-dots.js";
import { HalftoneCmyk, halftoneCmykPresets } from "./shaders/halftone-cmyk.js";
import { GemSmoke, gemSmokePresets } from "./shaders/gem-smoke.js";
import { isPaperShaderElement, getShaderColorFromString } from "@paper-design/shaders";
import {
  colorPanelsMeta,
  dotOrbitMeta,
  godRaysMeta,
  grainGradientMeta,
  meshGradientMeta,
  metaballsMeta,
  pulsingBorderMeta,
  simplexNoiseMeta,
  smokeRingMeta,
  swirlMeta,
  voronoiMeta,
  warpMeta,
  heatmapMeta,
  staticMeshGradientMeta,
  staticRadialGradientMeta
} from "@paper-design/shaders";
export {
  ColorPanels,
  Dithering,
  DotGrid,
  DotOrbit,
  FlutedGlass,
  GemSmoke,
  GodRays,
  GrainGradient,
  HalftoneCmyk,
  HalftoneDots,
  Heatmap,
  ImageDithering,
  LiquidMetal,
  MeshGradient,
  Metaballs,
  NeuroNoise,
  PaperTexture,
  PerlinNoise,
  PulsingBorder,
  ShaderMount,
  SimplexNoise,
  SmokeRing,
  Spiral,
  StaticMeshGradient,
  StaticRadialGradient,
  Swirl,
  Voronoi,
  Warp,
  Water,
  Waves,
  colorPanelsMeta,
  colorPanelsPresets,
  ditheringPresets,
  dotGridPresets,
  dotOrbitMeta,
  dotOrbitPresets,
  flutedGlassPresets,
  gemSmokePresets,
  getShaderColorFromString,
  godRaysMeta,
  godRaysPresets,
  grainGradientMeta,
  grainGradientPresets,
  halftoneCmykPresets,
  halftoneDotsPresets,
  heatmapMeta,
  heatmapPresets,
  imageDitheringPresets,
  isPaperShaderElement,
  liquidMetalPresets,
  meshGradientMeta,
  meshGradientPresets,
  metaballsMeta,
  metaballsPresets,
  neuroNoisePresets,
  paperTexturePresets,
  perlinNoisePresets,
  pulsingBorderMeta,
  pulsingBorderPresets,
  simplexNoiseMeta,
  simplexNoisePresets,
  smokeRingMeta,
  smokeRingPresets,
  spiralPresets,
  staticMeshGradientMeta,
  staticMeshGradientPresets,
  staticRadialGradientMeta,
  staticRadialGradientPresets,
  swirlMeta,
  swirlPresets,
  voronoiMeta,
  voronoiPresets,
  warpMeta,
  warpPresets,
  waterPresets,
  wavesPresets
};
