// @paper-design/shaders/dist/index.js
/* * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                    Paper Shaders                    *
 *       https://github.com/paper-design/shaders       *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * */

import { ShaderMount, isPaperShaderElement } from "./shader-mount.js";
import {
  defaultObjectSizing,
  defaultPatternSizing,
  ShaderFitOptions
} from "./shader-sizing.js";
import {
  meshGradientFragmentShader,
  meshGradientMeta
} from "./shaders/mesh-gradient.js";
import {
  smokeRingMeta,
  smokeRingFragmentShader
} from "./shaders/smoke-ring.js";
import { neuroNoiseFragmentShader } from "./shaders/neuro-noise.js";
import {
  dotOrbitMeta,
  dotOrbitFragmentShader
} from "./shaders/dot-orbit.js";
import {
  dotGridFragmentShader,
  DotGridShapes
} from "./shaders/dot-grid.js";
import {
  simplexNoiseMeta,
  simplexNoiseFragmentShader
} from "./shaders/simplex-noise.js";
import {
  metaballsMeta,
  metaballsFragmentShader
} from "./shaders/metaballs.js";
import { perlinNoiseFragmentShader } from "./shaders/perlin-noise.js";
import { voronoiMeta, voronoiFragmentShader } from "./shaders/voronoi.js";
import { wavesFragmentShader } from "./shaders/waves.js";
import {
  warpMeta,
  warpFragmentShader,
  WarpPatterns
} from "./shaders/warp.js";
import { godRaysMeta, godRaysFragmentShader } from "./shaders/god-rays.js";
import { spiralFragmentShader } from "./shaders/spiral.js";
import { swirlMeta, swirlFragmentShader } from "./shaders/swirl.js";
import {
  ditheringFragmentShader,
  DitheringShapes,
  DitheringTypes
} from "./shaders/dithering.js";
import {
  grainGradientFragmentShader,
  grainGradientMeta,
  GrainGradientShapes
} from "./shaders/grain-gradient.js";
import {
  pulsingBorderMeta,
  pulsingBorderFragmentShader,
  PulsingBorderAspectRatios
} from "./shaders/pulsing-border.js";
import {
  colorPanelsFragmentShader,
  colorPanelsMeta
} from "./shaders/color-panels.js";
import {
  staticMeshGradientFragmentShader,
  staticMeshGradientMeta
} from "./shaders/static-mesh-gradient.js";
import {
  staticRadialGradientFragmentShader,
  staticRadialGradientMeta
} from "./shaders/static-radial-gradient.js";
import {
  paperTextureFragmentShader
} from "./shaders/paper-texture.js";
import { waterFragmentShader } from "./shaders/water.js";
import {
  flutedGlassFragmentShader,
  GlassDistortionShapes,
  GlassGridShapes
} from "./shaders/fluted-glass.js";
import {
  imageDitheringFragmentShader
} from "./shaders/image-dithering.js";
import {
  heatmapMeta,
  heatmapFragmentShader,
  toProcessedHeatmap
} from "./shaders/heatmap.js";
import {
  liquidMetalFragmentShader,
  LiquidMetalShapes,
  toProcessedLiquidMetal
} from "./shaders/liquid-metal.js";
import {
  HalftoneDotsTypes,
  HalftoneDotsGrids,
  halftoneDotsFragmentShader
} from "./shaders/halftone-dots.js";
import {
  HalftoneCmykTypes,
  halftoneCmykFragmentShader
} from "./shaders/halftone-cmyk.js";
import {
  gemSmokeMeta,
  gemSmokeFragmentShader,
  toProcessedGemSmoke,
  GemSmokeShapes
} from "./shaders/gem-smoke.js";
import { getShaderColorFromString } from "./get-shader-color-from-string.js";
import { getShaderNoiseTexture } from "./get-shader-noise-texture.js";
import { emptyPixel } from "./empty-pixel.js";
export {
  DitheringShapes,
  DitheringTypes,
  DotGridShapes,
  GemSmokeShapes,
  GlassDistortionShapes,
  GlassGridShapes,
  GrainGradientShapes,
  HalftoneCmykTypes,
  HalftoneDotsGrids,
  HalftoneDotsTypes,
  LiquidMetalShapes,
  PulsingBorderAspectRatios,
  ShaderFitOptions,
  ShaderMount,
  WarpPatterns,
  colorPanelsFragmentShader,
  colorPanelsMeta,
  defaultObjectSizing,
  defaultPatternSizing,
  ditheringFragmentShader,
  dotGridFragmentShader,
  dotOrbitFragmentShader,
  dotOrbitMeta,
  emptyPixel,
  flutedGlassFragmentShader,
  gemSmokeFragmentShader,
  gemSmokeMeta,
  getShaderColorFromString,
  getShaderNoiseTexture,
  godRaysFragmentShader,
  godRaysMeta,
  grainGradientFragmentShader,
  grainGradientMeta,
  halftoneCmykFragmentShader,
  halftoneDotsFragmentShader,
  heatmapFragmentShader,
  heatmapMeta,
  imageDitheringFragmentShader,
  isPaperShaderElement,
  liquidMetalFragmentShader,
  meshGradientFragmentShader,
  meshGradientMeta,
  metaballsFragmentShader,
  metaballsMeta,
  neuroNoiseFragmentShader,
  paperTextureFragmentShader,
  perlinNoiseFragmentShader,
  pulsingBorderFragmentShader,
  pulsingBorderMeta,
  simplexNoiseFragmentShader,
  simplexNoiseMeta,
  smokeRingFragmentShader,
  smokeRingMeta,
  spiralFragmentShader,
  staticMeshGradientFragmentShader,
  staticMeshGradientMeta,
  staticRadialGradientFragmentShader,
  staticRadialGradientMeta,
  swirlFragmentShader,
  swirlMeta,
  toProcessedGemSmoke,
  toProcessedHeatmap,
  toProcessedLiquidMetal,
  voronoiFragmentShader,
  voronoiMeta,
  warpFragmentShader,
  warpMeta,
  waterFragmentShader,
  wavesFragmentShader
};
